Enable @QueryProjection annotation support for KSP.

a3magic3pocket opened the pull request:
fixes #996 

Could you please check this feature at @IceBlizz6 if possible?
Thank you.

IceBlizz6 commented:
Hey,
This looks like great, thank you!

I have just a few comments on this before we merge.

1)
In QueryDslProcessor you are using mapNotNull and you are returning null for invalid usages of these annotations.
So it seems like invalid usages are now ignored rather than alerting the user of misconfiguration.
I tend to prefer more explicitly alerting the user in these cases, i can't think of a case where use would want to use these annotations and not let them be picked up by this module.

I'm open for discussion on this one if you have thoughts on this, but my initial idea is to change mapNotNull into map and use throw/error instead of returning null.


2) 
```kotlin
is KSFunctionDeclaration -> {
    if (declaration.isConstructor()) {
        declaration.parent as? KSClassDeclaration
    } else {
        null
    }
}
```
Is this part only valid for QueryProjection?
Thinking maybe it should check if the annotation was QueryProjection, but looking over the code i must admit that i don't see an easy solution to that without making many further code changes so we may accept it as it is.


3)
So annotating constructors instead of the class is now possible as i understand it (?)
Would you be able to add some code into the example project that annotates the constructor?
This is both helpful to show others how it could be used, but also to ensure that it works as expected.

a3magic3pocket commented:
Hello @IceBlizz6,
Thank you for your kind comments.

Based on your feedback, I have modified the existing source code.
Please review it again.

- 1, 2
In the QueryDslProcessor, I have implemented logic to execute only when the @QueryProjection is used.
I also added a check to ensure it only runs when the @QueryProjection annotation is present on the constructor function and class.
In all other cases, I have made it throw an error to alert the user to the issue.

- 3
I have added an example showing how to apply the @QueryProjection annotation to a constructor function.

If my modifications are different from your intention or if you think there are any other areas that need adjustment, please feel free to let me know.

Thank you.

IceBlizz6 commented:
Thank you, the code changes looks good!
But i seem to be encountering a compile-time error when i try to run the tests in querydsl-example-ksp-codegen.

```
querydsl-example-ksp-codegen/src/test/kotlin/Tests.kt:131:25 Unresolved reference 'QPersonClassConstructorDTO'.
querydsl-example-ksp-codegen/src/test/kotlin/Tests.kt:149:25 Unresolved reference 'QPersonClassDTO'.
```

On my computer i need to add these lines to make it work.
```kotlin
import com.querydsl.example.ksp.QPersonClassConstructorDTO
import com.querydsl.example.ksp.QPersonClassDTO
```

It doesn't seem like github workflows has been set up to build this example project, so it won't be able to confirm whether this is an actual issue in the code or if this is a configuration issue on my local machine.

I just wanted to confirm with you that you are not encountering this issue?

a3magic3pocket commented:
Hello @IceBlizz6,

Thank you for your feedback.

After checking the issue you pointed out, I realized that it was my mistake.

I found that some import statements were missing when copying from the local Gradle querydsl-example environment to querydsl/querydsl-examples/querydsl-example-ksp-codegen.

I have added the missing import statements and pushed the changes.

Thank you!
